Add `directive` property to follow ESTree

mysticatea opened the pull request:
Recently, ESTree spec added `ExpressionStatement#directive` property for directive prologues: https://github.com/estree/estree/compare/b3de58c9997504d6fba04b72f76e6dd1619ee4eb...1da8e603237144f44710360f8feb7a9977e905e0

This PR makes Acorn following the spec change.

marijnh commented:
I find this implementation a bit verbose and fragile. Couldn't this be done as a post-processing pass on these bodies, rather than communicated through some piece of state? I.e. after a function body or top-level body is parsed, run a function on it that colors all the leading string literals as directives?

mysticatea commented:
Thank you for review.
OK, I will change the implementation. Merely I avoided extra looping.

marijnh commented:
That's true, but compared to the work that goes into building up a block, checking the type of a few nodes is probably not going to be noticeable.

mysticatea commented:
I have done.

mysticatea commented on src/loose/statement.js:
```diff
@@ -329,7 +330,8 @@ lp.parseFunction = function(node, isStatement, isAsync) {
   node.params = this.parseFunctionParams()
-  node.body = this.parseBlock()
+  this.parseFunctionBody(node)
+  this.adaptDirectivePrologue(node.body)
```
Ops, sorry, I must remove this line.

mysticatea replied:
Removed.

marijnh commented on src/loose/statement.js:
```diff
@@ -446,3 +448,19 @@ lp.parseExportSpecifierList = function() {
 }
+
+// Set `ExpressionStatement#directive` property for directive prologues.
+lp.adaptDirectivePrologue = function(statements) {
```
It seems like you could just call these in the normal parser here, rather than duplicating them.

mysticatea replied:
Oh, I have forgotten that `this.toks` is the normal parser.

marijnh commented on src/loose/statement.js:
```diff
@@ -329,7 +330,8 @@ lp.parseFunction = function(node, isStatement, isAsync) {
   this.inAsync = node.async
   node.params = this.parseFunctionParams()
-  node.body = this.parseBlock()
+  this.parseFunctionBody(node)
```
This seems to allow brace-less function bodies for regular functions — I don't think that logic needs to be in `parseFunctionBody`, since it only applies to arrow functions.

mysticatea replied:
`parseMethod` has the behavior, can parse the body as an expression, but can I remove it?

https://github.com/ternjs/acorn/pull/593/files#diff-d91a17a6a9019eba5785445f28c72421L518

marijnh replied:
Yeah, that doesn't seem to make a lot of sense

mysticatea commented:
I changed two:

- I removed duplicated LooseParser's functions: `lp.adaptDirectivePrologue` and `lp.isDirectiveCandidate`.
- I changed the code to parse body in `lp.parseFunction` to not use `this.parseFunctionBody(node)`.

mysticatea commented:
And:

- I removed brace-less function bodies from `lp.parseMethod`.

marijnh commented:
Great! Merged as bf94140

mysticatea commented:
Thank you!

RReverser commented on src/loose/expression.js:
```diff
@@ -514,8 +514,8 @@ lp.parseMethod = function(isGenerator, isAsync) {
     node.async = !!isAsync
   this.inAsync = node.async
   node.params = this.parseFunctionParams()
-  node.expression = this.options.ecmaVersion >= 6 && this.tok.type !== tt.braceL
```
Was this supposed to be removed? I don't see where `.expression` is set now.

RReverser replied:
Oh, I see, methods don't allow expressions anymore (they did in FF) so this code is obsolete indeed.
